ls1intum/Artemis
# `Lectures`: Allow instructors to filter lectures

### Checklist
#### General



#### Client

### Motivation and Context


This PR introduces a feature that allows instructors to filter their lectures based on the start and end dates.
Can be useful if the course has many lectures and the instructor wants to reduce clutter by filtering out past lectures or quickly find a lecture that is currently in progress to add an attachment.
It also allows instructors to find lectures for which the start or end date(s) are undefined.



### Description


Credit to @sunsetcoder at #5176 for the original code/idea.

Added a filter button with a dropdown menu and 4 options:
- Past: Filters lectures for which the end date is before the current local time.
- Current: Filters lectures for which the start date is same or before the current local time and the end date is after the current time (i.e., lectures currently in progress)
- Future: Filters lectures for which the start date is after the current local time.
- Unspecified Date(s): Filters Lectures for which the start and end date is undefined.

I added the unit tests (replaced the Cypress tests), resolved some merge conflicts, did some refactoring and fixed the filtering logic.

### Steps for Testing

Prerequisites:
- Instructor

1. Log in to Artemis
2. Navigate to Course Management page
3. Click on Lectures
4. Create lectures with start dates in the past, current, and future, and some without start/end dates specified.
5. Use the filter dropdown menu to ensure that the appropriate lectures are filtered out.
6. Ensure the functionality of the rest of the component remains unchanged (import, delete, edit still works)

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| lecture.component.ts | 88% | 96% |

### Screenshots



<img src="https://user-images.githubusercontent.com/33063132/171255035-24f6facd-51d2-4a7c-8a32-92937ccded3d.png" width="400">
